ons-spark/pyspark-intro/pyspark-intro.ipynb

```diff
@@ -0,0 +1,1550 @@
+    "\n",
+    "We will not aim to cover all the PySpark DataFrame functionality or go into detail of how Spark works, but instead focus on practicality by performing some common operations on an example dataset. There are a handful of exercises that you can complete while reading this article.\n",
+    "\n",
+    "Prerequisites for this article are some basic knowledge of Python and pandas. If you are completely new to Python then it is recommended to complete an introductory course first; your organisation may have specific Python training. Other resources include the [Python Tutorial](https://docs.python.org/3/tutorial/) and [10 Minutes to pandas](https://pandas.pydata.org/pandas-docs/stable/user_guide/10min.html).\n",
+    "\n",
+    "### PySpark: a quick introduction\n",
+    "\n",
+    "Although this article focusses on practical usage to enable you to quickly use PySpark, you do need to understand some basic theory of Spark and distributed computing.\n",
+    "\n",
+    "Spark is a powerful tool used to process huge data in an efficient way. We can access Spark in Python with the PySpark package. Spark has DataFrames, consisting of rows and columns, similar to pandas. Many of the operations are also similar, if not identically named: e.g. you can select and add columns, filter rows, group and aggregate.\n",
+    "\n",
+    "The key difference between PySpark and pandas is where the DataFrame is processed:\n",
+    "- pandas DataFrames are processed on the *driver*; this could be on a local machine using a desktop IDE such as Spyder or PyCharm, or on a server, e.g. in a dedicated Docker container (such as a CDSW session). The amount of data you can process is limited to the driver memory, so pandas is suitable for smaller data.\n",
+    "- PySpark DataFrames are processed on the *Spark cluster*. This is a big pool of linked machines, called *nodes*. You can have much greater memory capacity with Spark and so is suitable for big data.\n",
+    "\n",
+    "The DataFrame is also processed differently:\n",
+    "- In pandas, the DataFrame changes in memory at each point, e.g. you could create a DataFrame by reading from a CSV file, select some columns, filter the rows, add a column and then write the data out. With each operation, the DataFrame is physically changing in memory. This can be useful for debugging as it is easy to see intermediate outputs.\n",
+    "- In PySpark, DataFrames are *lazily evaluated*. We give Spark a set of instructions, called *transformations*, which are only evaluated when necessary, for instance to get a row count or write out data to a file, referred to as an *action*. In the example above, the plan is triggered once the data are set to write out to a file.\n",
+    "\n",
+    "It is recommended to read the Introduction to Spark Theory and Selecting the Right Package articles to help understand more about Spark.\n",
```

Comment:
Good question. We should have an article which has the key info from Using DAP, which I've provisionally called Introduction to Spark Theory but that's subject to change. The Selecting the Right Package is the tool selection article, not quite sure what name we agreed on for that

Reply:
I don't see us getting these in the book any time soon so it might be best to remove reference? Maybe just delete that sentence for now and I can add more links once these bits of guidance have been written. We could link to the Learning Spark textbook maybe? https://pages.databricks.com/rs/094-YMS-629/images/LearningSpark2.0.pdf?utm_medium=email&utm_source=databricks&utm_campaign=7014N0000026tqzQAA

Reply:
I've reworded this to link through to the whole Understanding and Optimising Spark chapter and put a link to the Databricks book there.